src/osd/OSDMap.cc

```diff
@@ -1957,12 +1957,19 @@ bool OSDMap::check_pg_upmaps(
     // okay, upmap is valid
     // continue to check if it is still necessary
     auto i = pg_upmap.find(pg);
-    if (i != pg_upmap.end() && raw == i->second) {
-      ldout(cct, 10) << " removing redundant pg_upmap "
-                     << i->first << " " << i->second
-                     << dendl;
-      to_cancel->push_back(pg);
-      continue;
+    if (i != pg_upmap.end()) {
+      if (i->second == raw) {
+        ldout(cct, 10) << "removing redundant pg_upmap " << i->first << " "
+                       << i->second << dendl;
+        to_cancel->push_back(pg);
+        continue;
+      }
+      if ((int)i->second.size() > get_pg_pool_size(pg)) {
```

Comment:
hmm, i think we should further narrow the condition down to **!=** to catch both oversize and undersize cases, e.g., we change pool size to a larger value after injecting some pg_upmaps into the osdmap.